// ngx-progress-bar/ngx-progress-bar.pure.amd.js
define("ngx-progress-bar/index", ["require", "exports", "@angular/core", "@angular/common"], function (require, exports, core_1, common_1) {
    var ProgressBar = (function () {
        function ProgressBar() {
            this.value = 0;
            this.max = 100;
        }
        ProgressBar.prototype.getProgress = function () {
            if (this.max === 0)
                return 0;
            return this.value / this.max * 100;
        };
        return ProgressBar;
    }());
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], ProgressBar.prototype, "value", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], ProgressBar.prototype, "max", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], ProgressBar.prototype, "title", void 0);
    ProgressBar = __decorate([
        core_1.Component({
            selector: "progress-bar",
            template: "\n<div class=\"progress\">\n    <div class=\"progress-bar\"\n        role=\"progressbar\"\n        [attr.aria-valuenow]=\"getProgress()\"\n        aria-valuemin=\"0\"\n        [attr.aria-valuemax]=\"getProgress()\" [ngStyle]=\"{ width: getProgress() + '%' }\">\n        <span [hidden]=\"!title\" class=\"progress-bar-title\">{{ title }}</span>\n        <span class=\"sr-only\">{{ getProgress() }}% Complete</span>\n    </div>\n</div>\n"
        })
    ], ProgressBar);
    exports.ProgressBar = ProgressBar;
    var ProgressBarModule = (function () {
        function ProgressBarModule() {
        }
        return ProgressBarModule;
    }());
    ProgressBarModule = __decorate([
        core_1.NgModule({
            imports: [
                common_1.CommonModule
            ],
            declarations: [
                ProgressBar
            ],
            exports: [
                ProgressBar
            ]
        })
    ], ProgressBarModule);
    exports.ProgressBarModule = ProgressBarModule;
});
define("ngx-progress-bar", ["require", "exports", "ngx-progress-bar/index"], function (require, exports, index_1) {
    function __export(m) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }
    __export(index_1);
});
